### PR TITLE
Set up Azure-CI

### LINKED ISSUE #993: Use Azure for CI?
https://azure.microsoft.com/en-us/blog/announcing-azure-pipelines-with-unlimited-ci-cd-minutes-for-open-source/
> Starting today, Azure Pipelines provides unlimited CI/CD minutes and 10 parallel jobs to every open source project for free. All open source projects run on the same infrastructure that our paying customers use.

via @ndmitchell

### PR DESCRIPTION
**Pull request by power-fungus**
closes #993. 

This is still a work in progress and I have setup one test-job. However, the job fails. See https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=24 

I need help fixing the test. What is needed to successfully run the tests?

### CODE REVIEW

**Comment by bubba** [comment]
Liquid Haskell needs to be installed for the functional tests, and IIRC in CircleCI it comes from the docker image

**Comment by power-fungus** [comment]
I have added `liquidhaskell` to the build, but it still fails. 
See: https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=31

@fendor @bubba, any ideas?

**Comment by bubba** [comment]
Those tests that are failing are probably caused by the Hoogle database not being generated beforehand. CodeActions need a `stack exec Hoogle generate` to be run beforehand in the CI job, in order to lookup which packages to import etc.

**Comment by power-fungus** [comment]
Will try.

**Comment by power-fungus** [comment]
Still no luck. The liquidhaskell-related tests still fail. See: https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=32

File: .azure/linux-stack.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    displayName: Build `hie`
+  - script: |
+      export PATH=$HOME/.local/bin:$PATH
+      stack build --test --bench --only-dependencies
+      stack install # `hie` binary required for tests
+      # TODO: get `cabal` binary from somewhere else?
+      stack install cabal-install
+      stack exec hoogle generate
+      cabal v1-update
+      cabal v1-install liquidhaskell --symlink-bindir=$HOME/.local/bin
```
Comment (bubba) [inline .azure/linux-stack.yml]: @power-fungus It looks like this is installing liquidhaskell 0.6.0.0, but the one we use for testing is 0.8.2.4 I believe. Can you try `cabal new-install liquidhaskell-0.8.2.4`? Although I'm not sure what versions of ghc this builds with

Comment (alanz) [reply .azure/linux-stack.yml]: You need GHC 8.2.2 to install liquidhaskell. I just rebuild my GHC stuff, and installed with

    cabal new-install liquidhaskell -w ghc-8.2.2

**Comment by power-fungus** [comment]
I am unable to get all liquidhaskell tests to work. I have tried your suggestions about `liquidhaskell-0.8.2.4` and `ghc-8.2.2` but the tests still fail. (https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=35)

Any other ideas?

**Comment by power-fungus** [comment]
Thanks to @fendor, the first build has just succeeded, see #1283.
Build: https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=38

Now, I will work on getting the build to work on all platforms with all ghc-versions

File: .azure/windows-stack.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      stack-8.2.2:
+        YAML_FILE: stack-8.2.2.yaml
+  steps:
+  - bash: |
+      git submodule sync
+      git submodule update --init
+    displayName: Sync submodules
+  - bash: |
+      export STACK_ROOT="$(Build.SourcesDirectory)"/.stack-root
+      curl -sSkL http://www.stackage.org/stack/windows-i386 -o /usr/bin/stack.zip
```
Comment (jneira) [inline .azure/windows-stack.yml]: Hi, why use windows-i386 instead x86_64?, i think the last one is the most used lately

**Comment by jneira** [comment]
It seems azure will get cache for the builds soon: https://github.com/Microsoft/azure-pipelines-yaml/pull/113
It will be great for haskell projects

File: .azure/windows-stack.yml (hunk 2 above)
Comment (jneira) [reply .azure/windows-stack.yml]: Even there is no ghc-8.6.4 for win32, it is strange

Comment (fendor) [reply .azure/windows-stack.yml]: It is strange indeed! Maybe we can raise an issue? or find an associated issue?

Comment (jneira) [reply .azure/windows-stack.yml]: Anyway i would use 64 bits arch by default, like in linux builds, it is the most common arch nowadays

**Comment by jneira** [comment]
I've request a pr to @power-fungus branch with changes that seems to make progresses in win builds (no the fail in the test step): https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=9

**Comment by power-fungus** [comment]
Thank you @jneira. Your changes are very valuable but as far I can tell, the builds still fail :(

Can someone investigate why the tests fail even on linux?

**Comment by jneira** [comment]
The pr build (https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=59) has failed cause the pr was merged and the build couldnt checkout the code
But the build of main branch will fail in windows cause my last untested commit breaks the PATH env var. Sorry for that, i'll make a new pr (testing a complete build before)

**Comment by jneira** [comment]
I've setup [another pr](https://github.com/power-fungus/haskell-ide-engine/pull/4) that make the windows builds reach the `run tests` step. 
The final test build will be [this](https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=49) but the same config was tested [here](https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=45)

**Comment by power-fungus** [comment]
For `8.6.2` and `8.6.1`, the installation of the ghc fails on Mac. Is this a known issue with mac and these ghc versions?

**Comment by power-fungus** [comment]
with help of @jneira, we have managed to get the build working on all platforms and almost-all ghc-versions: https://dev.azure.com/haskell-ide-engine/haskell-ide-engine-power-fungus/_build/results?buildId=76

There is a catch: no tests are executed, since they fail almost everywhere. I suggest merging this PR and adopting azure as build-check. Getting the tests to work there should a PR of its own.

**Comment by power-fungus** [comment]
I would consider this PR done as a first step.

**Comment by alanz** [comment]
As far as I can tell this looks good, happy for it to go in as a base build.

**Comment by power-fungus** [comment]
I'll merge this PR since this the feedback has been positive.

**Comment by power-fungus** [comment]
I tried officially add this repository for a new pipeline. But I have not the access-rights to change the hooks of this repository.

There are 2 options: 
* I or @fendor get (temporary) access rights for hooks to this repository.
* I invite some one with the appropriate access rights to the `haskell-ide-engine` team on azure-ci and that person adds the pipeline

**Pull request merged**